src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -808,6 +812,102 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
+            {
+                // `UnmanagedCallersOnly` can only be applied to ordinary static methods or local functions.
+                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, arguments.AttributeSyntaxOpt.Location);
+            }
+
+            if (!arguments.Attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            {
+                foreach (var (key, value) in arguments.Attribute.CommonNamedArguments)
+                {
+                    if (key != "CallConvs"
+                        || value.Kind != TypedConstantKind.Array
+                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    {
+                        continue;
+                    }
+
+                    foreach (var callConvTypedConstant in value.Values)
+                    {
+                        Debug.Assert(callConvTypedConstant.Kind == TypedConstantKind.Type);
+                        if (!(callConvTypedConstant.ValueInternal is NamedTypeSymbol callConvType)
+                            || !FunctionPointerTypeSymbol.IsCallingConventionModifier(callConvType))
+                        {
+                            // `{0}` is not a valid calling convention type for 'UnmanagedCallersOnly'.
+                            arguments.Diagnostics.Add(ErrorCode.ERR_InvalidUnmanagedCallersOnlyCallConv, arguments.AttributeSyntaxOpt.Location, callConvTypedConstant.ValueInternal ?? "null");
+                        }
+                    }
+                }
+            }
+
+            var (returnTypeSyntax, genericTypeParameters) = SyntaxNode switch
+            {
+                MethodDeclarationSyntax m => (m.ReturnType, m.TypeParameterList),
+                LocalFunctionStatementSyntax l => (l.ReturnType, l.TypeParameterList),
+                _ => default
+            };
+
+            if (returnTypeSyntax is null)
+            {
+                // We already issued an error for these above. We don't immediately bail out so we can error on invalid
+                // calling convention types as well, but erroring on parameter or return types will likely just be noise.
+                Debug.Assert(MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction));
+                return;
+            }
+
+            if (IsGenericMethod)
+            {
+                Debug.Assert(genericTypeParameters is not null);
+                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyMethodCannotBeGeneric, genericTypeParameters.GetLocation());
+            }
+
+            var containingType = ContainingType;
+            while (containingType is not null)
+            {
+                if (containingType.IsGenericType)
+                {
+                    arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyMethodCannotBeInGenericType, arguments.AttributeSyntaxOpt.Name.Location);
```

Comment:
>ERR_UnmanagedCallersOnlyMethodCannotBeInGenericType [](start = 56, length = 51)

Please combine with `ERR_UnmanagedCallersOnlyMethodCannotBeGeneric`.

And we should only report one error if the method and containing type are both generic. Please test that case if it is not already tested.